add deprecation chapter (#1850)
plone/documentation: docs/developer-guide/deprecation.md

)

def some_function_which_is_not_touched_at_all():
    pass
```

### Deprecate a GenericSetup profile

In GenericSetup, the `post_handler` attribute in ZCML can be used to call a function after the profile was applied.
Use this feature to issue a warning.

First, register the same profile twice, under both the new name and old.

```xml
<genericsetup:registerProfile
    name="default"
    title="My Fancy Package"
    directory="profiles/default"
    description="..."
    provides="Products.GenericSetup.interfaces.EXTENSION"
    />

<genericsetup:registerProfile
    name="some_confusing_name"
    title="My Fancy Package (deprecated)"
    directory="profiles/some_confusing_name"
    description="... (use profile default instead)"
    provides="Products.GenericSetup.interfaces.EXTENSION"
    post_handler=".setuphandlers.deprecate_profile_some_confusing_name"
    />
```

Then in {file}`setuphandlers.py`, add a function.

```python
import warnings

def deprecate_profile_some_confusing_name(tool):
    warnings.warn(
        'The profile with id "some_confusing_name" was renamed to "default".',
        DeprecationWarning
    )
```

> ### Deprecate a template position

Reviewer: ```suggestion
### Deprecate a template location
```

Reviewer 2: There's also a mixture of "position" and "location" in the narrative text. Does a template have a position or a location? Whichever is more appropriate, we should use in the narrative text as well.

For the record, location pertains to file system, whereas position pertains to an order within a group of similar items.

Reviewer 2: @plone/classicui-team please respond. Does a template have a "position" or a "location"? I'll update the text accordingly. Let's get this merged! Thank you!

Sometimes you need to move templates to new locations.
Since add-ons often use [`z3c.jbot`](https://github.com/zopefoundation/z3c.jbot) to override templates by their location, you'll need to point them to the new location as well as make sure that the override still works with the old location.

To deprecate a template, follow these steps.

1.  In the old package folder's {file}`__init__.py`, add a dictionary `jbot_deprecations` that maps the old template locations to their new counterparts.

    ```python
    jbot_deprecations = {
        "plone.locking.browser.info.pt": "plone.app.layout.viewlets.info.pt"
    }
    ```

1.  Add this deprecation snippet to the package {file}`configure.zcml` file.

    ```{code-block} xml

>     :emphasize-lines: 7-14

Reviewer 3: ```suggestion
    :emphasize-lines: 7-15
```

    description="... (use profile default instaed)"
    provides="Products.GenericSetup.interfaces.EXTENSION"
    post_handler=".setuphandlers.deprecate_profile_some_confusing_name"
    />
```

And in `setuphandlers.py` add a function:

```python
import warnings

def deprecate_profile_some_confusing_name(tool):
    warnings.warn(
        'The profile with id "some_confusing_name" was renamed to "default".',
        DeprecationWarning
    )
```

### Deprecating a template position

Sometimes we need to move templates to new locations. Since addons often use [z3c.jbot](https://github.com/zopefoundation/z3c.jbot) to override templates by their position, we need to point them to the new position as well as make sure that the override still works with the old position.

To deprecate a package:

1. In the old package folders `__init__.py` add a dictionary `jbot_deprecations` that maps the old template locations to their new counterparts, e.g.:

```python
jbot_deprecations = {
    "plone.locking.browser.info.pt": "plone.app.layout.viewlets.info.pt"
}
```

2. Add this deprecation snippet to the package `configure.zcml` file:

```{code-block} xml

> :emphasize-lines: 6,9-12

Reviewer 3: See https://plone6--1850.org.readthedocs.build/developer-guide/deprecation.html#deprecating-a-template-position:

<img width="1007" height="403" alt="Image" src="https://github.com/user-attachments/assets/79ed25c2-9b06-4480-b227-203a2f255755" />

```suggestion
:emphasize-lines: 7-14
```

---
myst:
  html_meta:
    "description": "How to implement deprecations in Plone, including Python, ZCML and templates."
    "property=og:description": "How to implement deprecations in Plone, including Python, ZCML and templates."
    "property=og:title": "Implement deprecations"
    "keywords": "deprecation, zcml, template, jbot, Plone, Python"
---

(developer-deprecation-label)=

# Deprecate code

This chapter describes how to enable deprecation warnings and best practices for deprecating code in Plone, Zope, and Python.

```{seealso}
For background on deprecation philosophy and use cases, see {doc}`/conceptual-guides/deprecation`.
```

## Enable deprecation warnings

This section describes how to enable deprecation warnings in Zope, Python, and tests.

### Zope

> Zope configures logging and warnings, so the steps as described below in {ref}`deprecation-warning-python-label` aren't needed.

Reviewer 2: If the steps in the next section aren't needed, then why do we include them?

Reviewer 4: With next step you refer to `debug-mode on`? IIRC it is needed to show the messages. I wrote this down years ago, so this may have been subject to change.

Reviewer 4: And with `cookiecutter-zope-instance` based installs we only need to use the PYTHONWARNINGS variable or start the interpreter with parameters (like below in Python).

Thinking further, I am not 100% sure if the above method is anyway ZServer only?
Does anybody in here know more details?

Reviewer 2: @jensens you might be able to see it more clearly on the pull request preview, beginning at https://plone6--1850.org.readthedocs.build/developer-guide/deprecation.html#zope.

This sentence refers to the next section entitled [Python](https://plone6--1850.org.readthedocs.build/developer-guide/deprecation.html#deprecation-warning-python-label), and it states that the steps in that section are not needed.

Reviewer 4: We can drop the whole section `Zope` with plone.recipe.zopeinstance since this feature was dropped with WSGI server usage. So, it does not make sense to keep the Python subheading, right?
See last bullet point at https://github.com/plone/plone.recipe.zope2instance/blob/0452e05386708564b17ae1de1aafe708a4d88ccb/CHANGES.rst#610-2018-12-28 

Shall we mention somehow this change in the developer guide or focus on the current state?

Reviewer 2: I wish I could offer advice, but this was before my time. I'll go with what you think is true and accurate, and to the best of your knowledge. We can also wait a few days for other old timers to offer advice. The only thing I can refer to is this search result:

https://6.docs.plone.org/search.html?q=plone.recipe.zopeinstance

I'd focus on the current state of Plone 6.x.x. I'd use a `{versionchanged}` or `{versionremoved}` MyST directive, but we should use Plone 6.x.x or 5.x.x as the referred version, whenever the package got released with Plone, not the package's version. The content of that directive should include whatever is helpful to Plone 6 developers for what has changed. There might even be some references in the [Upgrade Guide](https://6.docs.plone.org/backend/upgrading/index.html). Why, yes, there is: https://6.docs.plone.org/backend/upgrading/version-specific-migration/upgrade-to-52.html#wsgi

Does that help at all?

Reviewer 5: > We can drop the whole section `Zope` with plone.recipe.zopeinstance since this feature was dropped with WSGI server usage. So, it does not make sense to keep the Python subheading, right?

I think you mean we don't need to keep the *Zope* subheading.  The *Python* subheading still seems accurate.

I must say I *always* get deprecation warnings when starting Plone though, even though I don't think I have configured anything to enable it.  So when I don't want them, I explicitly *disable* them with `export PYTHONWARNINGS=ignore`.  This may depend on how Python is installed, and with which flags it was built.

> Shall we mention somehow this change in the developer guide or focus on the current state?

Indeed focus on current state, but if possible mention the change as Steve says.

Reviewer 4: We then have one subheading, that is a bit pointless.  But if it helps, ok.

Reviewer 2: There would be two, one for Python and the other for running tests.

Using `plone.recipe.zope2instance`, add the option `deprecation-warnings = on` to the buildout's `[instance]` section.

```cfg
[buildout]
parts = instance

[instance]
recipe = plone.recipe.zope2instance
# …
deprecation-warnings = on
# …
```

This adds the following line to the {file}`zope.conf` file.

```cfg
debug-mode on
```

Without the recipe, this can be set manually as well.
In {file}`zope.conf`, define custom filters for warnings, such as the following example.

```xml
<warnfilter>
    action always
    category exceptions.DeprecationWarning
</warnfilter>
```

(deprecation-warning-python-label)=

### Python

Enable warnings

:   Warnings are written to `stderr` by default, but `DeprecationWarning` output is surpressed by default.
    
    Output can be enabled by starting the Python interpreter with the {ref}`-W[all|module|once]  <python:using-on-warnings>` argument.
 

>     It's possible to enable output in code, too.

Reviewer 4: We are missing an important way of control here:

```suggestion
    Output can be enabled by starting the Python interpreter with the {ref}`-W[all|module|once]  <python:using-on-warnings>` argument.
    
    As an alternative, the environment variable {envvar}`python:PYTHONWARNINGS` can be set to `default`, in other words, `PYTHONWANRINGS=default`.
 
    It's possible to enable output in code, too.
```

Reviewer 2: Amended to use a link to the Python object.

```

## Enable deprecation warnings

This section describes how to enable deprecation warnings in Zope, Python, and tests.

### Zope

Zope configures logging and warnings, so the steps as described below in {ref}`deprecation-warning-python-label` aren't needed.

Using `plone.recipe.zope2instance`, add the option `deprecation-warnings = on` to the buildout's `[instance]` section.

```cfg
[buildout]
parts = instance

[instance]
recipe = plone.recipe.zope2instance
# …
deprecation-warnings = on
# …
```

This adds the following line to the {file}`zope.conf` file.

```cfg
debug-mode on
```

Without the recipe, this can be set manually as well.
In {file}`zope.conf`, define custom filters for warnings, such as the following example.

```xml
<warnfilter>
    action always
    category exceptions.DeprecationWarning
</warnfilter>
```

(deprecation-warning-python-label)=

### Python

Enable warnings

:   Warnings are written to `stderr` by default, but `DeprecationWarning` output is surpressed by default.
    
    Output can be enabled by starting the Python interpreter with the {ref}`-W[all|module|once]  <python:using-on-warnings>` argument.
    

>     As an alternative, the environment variable {envvar}`python:PYTHONWARNINGS` can be set to `default`, in other words, `PYTHONWANRINGS=default`.

Reviewer 5: typo

```suggestion
    As an alternative, the environment variable {envvar}`python:PYTHONWARNINGS` can be set to `default`, in other words, `PYTHONWARNINGS=default`.
```

To enable deprecation warnings, use the `-W` command.

Given you're using a modern buildout with a virtual environment as recommended, the command would be the following

```shell
./bin/python -W module ./bin/test
```

## Deprecation best practices

It's recommended to follow these best practices when deprecating code.

### Vanilla deprecation messages

Python offers a built-in exception {exc}`DeprecationWarning` which can be issued using the standard library's {mod}`warnings` module.

Its basic usage is the following example.

```python
import warnings
warnings.warn("deprecated", DeprecationWarning)
```

### Move an entire module

Given a package {file}`old.pkg` with a module {file}`foo.py`, to move it to a package {file}`new.pkg` as {file}`bar.py`, go through the following steps.

[`zope.deprecation` Moving modules](https://zopedeprecation.readthedocs.io/en/latest/api.html#moving-modules) offers a helper.

1.  Move the {file}`foo.py` as {file}`bar.py` to the {file}`new.pkg`.
1.  At the old place, create a new {file}`foo.py`, and add to it the following lines of code.

    ```python
    from zope.deprecation import moved
    moved("new.pkg.bar", "Version 2.0")
    ```

> 1.  Now you can still import the namespace from `bar` at the old place, but get a deprecation warning.

Reviewer 5: This confuses me a bit.

```suggestion
1.  Now you can still import the module at the old place, but get a deprecation warning.
```

```xml
<genericsetup:registerProfile
    name="default"
    title="My Fancy Package"
    directory="profiles/default"
    description="..."
    provides="Products.GenericSetup.interfaces.EXTENSION"
    />

<genericsetup:registerProfile
    name="some_confusing_name"
    title="My Fancy Package (deprecated)"
    directory="profiles/some_confusing_name"
    description="... (use profile default instead)"
    provides="Products.GenericSetup.interfaces.EXTENSION"
    post_handler=".setuphandlers.deprecate_profile_some_confusing_name"
    />
```

Then in {file}`setuphandlers.py`, add a function.

```python
import warnings

def deprecate_profile_some_confusing_name(tool):
    warnings.warn(
        'The profile with id "some_confusing_name" was renamed to "default".',
        DeprecationWarning
    )
```

### Deprecate a template position

Sometimes you need to move templates to new locations.
Since add-ons often use [`z3c.jbot`](https://github.com/zopefoundation/z3c.jbot) to override templates by their position, you'll need to point them to the new position as well as make sure that the override still works with the old position.

> To deprecate a package, follow these steps.

Reviewer 5: ```suggestion
To deprecate a template, follow these steps.
```

1.  In the old package folder's {file}`__init__.py`, add a dictionary `jbot_deprecations` that maps the old template locations to their new counterparts.

    ```python
    jbot_deprecations = {
        "plone.locking.browser.info.pt": "plone.app.layout.viewlets.info.pt"
    }
    ```

1.  Add this deprecation snippet to the package {file}`configure.zcml` file.

    ```{code-block} xml
    :emphasize-lines: 7-14
    :linenos:
    
    <configure
        xmlns="http://namespaces.zope.org/zope"
        xmlns:browser="http://namespaces.zope.org/browser"
        xmlns:zcml="http://namespaces.zope.org/zcml"
        >
    
      <include

>           zcml:condition="installed z3c.jbot"
>           package="z3c.jbot"

Reviewer 5: We should explicitly load `meta.zcml` of z3c.jbot`:

```suggestion
      <include
          package="z3c.jbot"
          file="meta.zcml"
          zcml:condition="installed z3c.jbot"
```

That is the file that defines the `jbot-deprecations` feature and the `jbotDeprecated` directive.
This is how [`plone.locking` does it](https://github.com/plone/plone.locking/blob/4.0.0a3/src/plone/locking/configure.zcml#L14-L23).

---
myst:
  html_meta:
    "description": "How to implement deprecations in Plone, including Python, ZCML and templates."
    "property=og:description": "How to implement deprecations in Plone, including Python, ZCML and templates."
    "property=og:title": "Implement deprecations"
    "keywords": "deprecation, zcml, template, jbot, Plone, Python"
---

(developer-deprecation-label)=

# Deprecate code

This chapter describes how to enable deprecation warnings and best practices for deprecating code in Plone, Zope, and Python.

```{seealso}
For background on deprecation philosophy and use cases, see {doc}`/conceptual-guides/deprecation`.
```

## Enable deprecation warnings

> This section describes how to enable deprecation warnings in Zope, Python, and tests.

Reviewer 2: ```suggestion
This section describes how to enable deprecation warnings in Python, both in an interpreter and code, and when running tests.
```

### Zope

Zope configures logging and warnings, so the steps as described below in {ref}`deprecation-warning-python-label` aren't needed.

Using `plone.recipe.zope2instance`, add the option `deprecation-warnings = on` to the buildout's `[instance]` section.

```cfg
[buildout]
parts = instance

[instance]
recipe = plone.recipe.zope2instance
# …
deprecation-warnings = on
# …
```

This adds the following line to the {file}`zope.conf` file.

```cfg
debug-mode on
```

Without the recipe, this can be set manually as well.
In {file}`zope.conf`, define custom filters for warnings, such as the following example.

```xml
<warnfilter>
    action always
    category exceptions.DeprecationWarning
</warnfilter>

> ```

Reviewer 2: Per @mauritsvanrees's suggestion.

```suggestion
```

(deprecation-warning-python-label)=

### Python

Enable warnings

:   Warnings are written to `stderr` by default, but `DeprecationWarning` output is surpressed by default.
    

>     Output can be enabled by starting the Python interpreter with the {ref}`-W[all|module|once]  <python:using-on-warnings>` argument.

Reviewer 2: ```suggestion
    Output can be enabled by starting the Python interpreter with the {ref}`-W[all|module|once] <python:using-on-warnings>` argument.
```

    
    As an alternative, the environment variable {envvar}`python:PYTHONWARNINGS` can be set to `default`, in other words, `PYTHONWARNINGS=default`.
 
    It's possible to enable output in code, too.
    
    ```python
    import warnings
    warnings.simplefilter("module")
    ```

Configure logging

:   Once output is enabled, it's possible to use {func}`python:logging.captureWarnings` to redirect warnings to the logger.

    ```python
    import logging
    logging.captureWarnings(True)
    ```

### Running tests

In Plone, test deprecation warnings are not shown by default.
The {file}`zope.conf` setting is not taken into account.

To enable deprecation warnings, use the `-W` command.

> Given you're using a modern buildout with a virtual environment as recommended, the command would be the following

Reviewer 2: ```suggestion
Given you're using a modern buildout with a virtual environment as recommended, the command would be the following.
```
